File: lib/protocol/remote_session.dart
import 'dart:async';

import 'package:flutter/foundation.dart';

import 'channel_client.dart';
import 'constants.dart';
import 'link_params.dart';
import 'relay_client.dart';
import 'rpc_frames.dart';

/// High-level facade: relay connect -> pair -> bootstrap -> workspace bridge.
class RemoteSession {
  final LinkParams params;
  final void Function(String line)? onLog;

  late final RelayClient relay;
  final _matchers = <String, bool Function(Map<String, dynamic>)>{};
  final _completers = <String, Completer<Map<String, dynamic>>>{};
  StreamSubscription? _payloadSub;

  final _workspaceListUpdated = StreamController<dynamic>.broadcast();
  Stream<dynamic> get workspaceListUpdated => _workspaceListUpdated.stream;

  RemoteSession(this.params, {this.onLog}) {
    relay = RelayClient(params, onLog: onLog);
    _payloadSub = relay.payloads.listen(_dispatch);
    relay.stateListenable.addListener(_onRelayState);
  }

  void _log(String line) => onLog?.call(line);

  Future<void> connect() => relay.start();

  final _activeBridges = <Bridge>[];
  bool _needsBridgeRecovery = false;

  /// 断线后重新配对成功的回调（每次真实重连恰好触发一次）。
  /// 上层（ZApp）在这里做状态对账：清本地相位覆盖、强制重同步订阅、
  /// 重拉任务列表——桌面端可能已崩溃重启，一切以服务端为准（BUG-34）。
  void Function()? onRePaired;

  void _onRelayState() {
    final s = relay.state;
    if (s == RelayState.reconnecting || s == RelayState.error) {
      if (_activeBridges.isNotEmpty) {
        _needsBridgeRecovery = true;
        for (final b in _activeBridges) {
          b.degraded.value ??= 'reconnecting';
        }
      }
      return;
    }
    if (s == RelayState.paired && _needsBridgeRecovery) {
      _needsBridgeRecovery = false;
      for (final b in List<Bridge>.from(_activeBridges)) {
        unawaited(b.recoverWithRetry());
      }
      // 桥恢复只是链路层；数据层的对账交给上层（清覆盖/重同步/重拉列表）。
      unawaited(Future.sync(() => onRePaired?.call()));
    }
  }

  /// Waits until the relay is paired with the desktop.
  Future<void> waitPaired({Duration timeout = const Duration(seconds: 60)}) {
    if (relay.state == RelayState.paired) return Future.value();
    final completer = Completer<void>();
    Timer? timer;
    void listener() {
      if (relay.state == RelayState.paired && !completer.isCompleted) {
        timer?.cancel();
        completer.complete();
      }
    }

    relay.stateListenable.addListener(listener);
    timer = Timer(timeout, () {
      if (!completer.isCompleted) completer.completeError(TimeoutException('配对超时'));
    });
    return completer.future.whenComplete(() {
      timer?.cancel();
      relay.stateListenable.removeListener(listener);
    });
  }

  void _dispatch(Map<String, dynamic> payload) {
    final type = payload['zcode_type'];
    if (type == pushWorkspaceListUpdated) {
      _workspaceListUpdated.add(payload['result']);
      return;
    }
    if (type == pushBridgeDegraded) {
      final id = payload['bridgeSessionId'] as String?;
      final reason = '${payload['reason'] ?? 'unknown'}';
      _log('[bridge] degraded: $id reason=$reason');
      for (final b in _activeBridges) {
        if (b.info['bridgeSessionId'] == id) {
          b.degraded.value = reason;
          unawaited(b.recoverWithRetry());
        }
      }
      return;
    }
    if (type == 'rpc-frame' || type == 'rpc-frame-ack') {
      final id = payload['bridgeSessionId'] as String?;
      final bridge = id == null ? null : _frameRouters[id];
      if (bridge != null) {
        bridge.frames.accept(payload);
      } else if (id != null) {
        (_pendingBridgePayloads[id] ??= []).add(payload);
      }
      return;
    }
    // Signaling request/response: matchers keyed by requestId. Responses are
    // not guaranteed to echo the id, so every matcher gets a chance.
    final done = <String>[];
    _matchers.forEach((requestId, match) {
      final completer = _completers[requestId];
      if (completer != null && !completer.isCompleted && match(payload)) {
        done.add(requestId);
        completer.complete(payload);
      }
    });
    for (final id in done) {
      _matchers.remove(id);
      _completers.remove(id);
    }
  }

  Future<Map<String, dynamic>> request(
    Map<String, dynamic> payload,
    bool Function(Map<String, dynamic>) match, {
    Duration timeout = const Duration(seconds: 30),
  }) {
    final requestId = payload['requestId'] as String;
    final completer = Completer<Map<String, dynamic>>();
    _matchers[requestId] = match;
    _completers[requestId] = completer;
    relay.sendPayload(payload);
    return completer.future.timeout(timeout, onTimeout: () {
      _matchers.remove(requestId);
      _completers.remove(requestId);
      throw TimeoutException('request $requestId timed out');
    });
  }

  /// bootstrap-request → bootstrap-response（任务列表总览）。
  Future<Map<String, dynamic>> bootstrap() async {
    final res = await request(
      {'zcode_type': sigBootstrap, 'requestId': genId('boot')},
      (p) => p['zcode_type'] == sigBootstrapResponse,
    );
    return (res['result'] as Map?)?.cast<String, dynamic>() ?? res;
  }

  Future<void> sendViewState({
    required String workspaceKey,
    String? taskId,
  }) {
    relay.sendPayload({
      'zcode_type': sigViewStateUpdate,
      'viewState': {
        'activeWorkspaceKey': workspaceKey,
        'activeTaskId': ?taskId,
        'updatedAt': DateTime.now().millisecondsSinceEpoch,
      },
      'deviceInfo': {
        'platform': platformName,
        'version': params.appVersion ?? 'zremote',
        'name': 'zremote',
      },
    });
    return Future.value();
  }

  final _frameRouters = <String, Bridge>{};
  final _pendingBridgePayloads = <String, List<Map<String, dynamic>>>{};
  int _bridgeGeneration = 0;

  /// workspace-bridge-open → workspace-bridge-ready → rpc-frame/IPC stack.
  Future<Bridge> openBridge(
    String workspaceKey, {
    String? taskId,
    Duration timeout = const Duration(seconds: 30),
  }) async {
    final requestedId = genId('bridge');
    final generation = ++_bridgeGeneration;
    final res = await request(
      {
        'zcode_type': sigBridgeOpen,
        'requestId': genId('open'),
        'bridgeSessionId': requestedId,
        'bridgeGeneration': generation,
        'workspaceKey': workspaceKey,
        'taskId': ?taskId,
      },
      (p) =>
          (p['zcode_type'] == sigBridgeReady ||
              p['zcode_type'] == sigBridgeError) &&
          p['bridgeSessionId'] == requestedId,
      timeout: timeout,
    );
    if (res['zcode_type'] == sigBridgeError) {
      throw StateError('workspace-bridge-error: ${res['error'] ?? res}');
    }
    final info =
        (res['bridge'] as Map?)?.cast<String, dynamic>() ?? <String, dynamic>{};
    _log('[bridge] ready: $info');
    final bridge = Bridge._(session: this, info: info);
    _attachStack(bridge, info);
    _activeBridges.add(bridge);
    final key = (info['workspaceKey'] as String?) ?? workspaceKey;
    unawaited(
        sendViewState(workspaceKey: key, taskId: bridge.initialTaskId ?? taskId));
    return bridge;
  }

  void _attachStack(Bridge bridge, Map<String, dynamic> info) {
    bridge._swapStack(info, _buildFrames(bridge, info));
    final id = info['bridgeSessionId'] as String? ?? '';
    _frameRouters[id] = bridge;
    final pending = _pendingBridgePayloads.remove(id);
    if (pending != null) {
      for (final payload in pending) {
        bridge.frames.accept(payload);
      }
    }
  }

  RpcFrames _buildFrames(Bridge bridge, Map<String, dynamic> info) {
    return RpcFrames(
      bridgeSessionId: '${info['bridgeSessionId'] ?? ''}',
      bridgeGeneration: (info['bridgeGeneration'] as num?)?.toInt(),
      recoveryId: info['recoveryId'] as String?,
      send: relay.sendPayload,
      onMessage: (bytes) => bridge.channels.handleMessage(bytes),
      onLog: onLog,
    );
  }

  Future<void> pokeRelay() async {
    relay.poke();
  }

  Future<void> dispose() async {
    relay.stateListenable.removeListener(_onRelayState);
    await _payloadSub?.cancel();
    for (final b in List<Bridge>.from(_activeBridges)) {
      b.dispose();
    }
    await relay.dispose();
    await _workspaceListUpdated.close();
  }
}

/// One workspace bridge: rpc-frame transport + channel client.
class Bridge {
  final RemoteSession session;
  Map<String, dynamic> info;
  late RpcFrames _frames;
  late ChannelClient channels;

  /// 转发 rpc-frame 分片入口（信令层按 bridgeSessionId 路由用）。
  RpcFrames get frames => _frames;

  /// Non-null while degraded (reconnecting / rpc-transport-fault).
  final ValueNotifier<String?> degraded = ValueNotifier(null);

  /// Bumped when the stack is swapped after recovery; subscriptions must
  /// resubscribe since server-side state died with the old bridge.
  final ValueNotifier<int> recovered = ValueNotifier(0);

  bool _disposed = false;

  Bridge._({required this.session, required this.info}) {
    _frames = RpcFrames(
      bridgeSessionId: '${info['bridgeSessionId'] ?? ''}',
      send: session.relay.sendPayload,
      onMessage: (bytes) => channels.handleMessage(bytes),
      onLog: session.onLog,
    );
    channels = ChannelClient(sendBody: _frames.sendMessage, onLog: session.onLog);
  }

  void _swapStack(Map<String, dynamic> newInfo, RpcFrames frames) {
    final old = _frames;
    _frames = frames;
    info = newInfo;
    channels = ChannelClient(sendBody: frames.sendMessage, onLog: session.onLog);
    old.dispose();
  }

  String? get workspaceKey => info['workspaceKey'] as String?;
  String? get initialTaskId => info['initialTaskId'] as String?;

  Map<String, dynamic> get scope => {
        'workspacePath': info['workspacePath'] ?? workspaceKey,
        if (info['workspaceIdentity'] != null)
          'workspaceIdentity': info['workspaceIdentity'],
      };

  /// Resolves once healthy, or throws [TimeoutException].
  Future<void> waitHealthy({Duration timeout = const Duration(seconds: 45)}) {
    if (_disposed || degraded.value == null) return Future.value();
    final completer = Completer<void>();
    void check() {
      if (degraded.value == null && !completer.isCompleted) completer.complete();
    }

    degraded.addListener(check);
    check();
    return completer.future.timeout(timeout, onTimeout: () {
      degraded.removeListener(check);
      throw TimeoutException('bridge 恢复超时: ${degraded.value}');
    }).whenComplete(() => degraded.removeListener(check));
  }

  /// Retries recovery until the bridge is healthy again.
  Future<void> recoverWithRetry() async {
    if (_disposed) return;
    for (var attempt = 1; attempt <= 8; attempt++) {
      if (_disposed) return;
      if (await _recoverOnce()) return;
      if (_disposed) return;
      session.onLog?.call('[bridge] recovery attempt $attempt failed');
      await Future<void>.delayed(const Duration(seconds: 3));
    }
  }

  Future<bool> _recoverOnce() async {
    final key = workspaceKey;
    if (key == null) {
      degraded.value = null;
      return true;
    }
    degraded.value = 'recovering';
    // 1) cheap path: workspace-reconnect-request
    try {
      final res = await session.request(
        {
          'zcode_type': sigBridgeReconnect,
          'requestId': genId('reconn'),
          'workspaceKey': key,
        },
        (p) =>
            p['zcode_type'] == sigBridgeReconnectResponse &&
            p['workspaceKey'] == key,
        timeout: const Duration(seconds: 15),
      );
      if (res['success'] == true) {
        degraded.value = null;
        recovered.value += 1;
        return true;
      }
    } on Object {
      // fall through to reopen
    }
    // 2) full reopen with recoveryId
    final requestedId = genId('bridge');
    final generation = ++session._bridgeGeneration;
    try {
      final res = await session.request(
        {
          'zcode_type': sigBridgeOpen,
          'requestId': genId('reopen'),
          'bridgeSessionId': requestedId,
          'bridgeGeneration': generation,
          if (info['recoveryId'] != null) 'recoveryId': info['recoveryId'],
          'workspaceKey': key,
        },
        (p) =>
            (p['zcode_type'] == sigBridgeReady ||
                p['zcode_type'] == sigBridgeError) &&
            p['bridgeSessionId'] == requestedId,
        timeout: const Duration(seconds: 30),
      );
      if (res['zcode_type'] == sigBridgeError) {
        throw StateError('${res['error'] ?? res}');
      }
      final newInfo =
          (res['bridge'] as Map?)?.cast<String, dynamic>() ?? <String, dynamic>{};
      final oldRouterId = '${info['bridgeSessionId'] ?? ''}';
      session._frameRouters.remove(oldRouterId);
      session._attachStack(this, newInfo);
      degraded.value = null;
      recovered.value += 1;
      session.onLog?.call('[bridge] reopened $key');
      return true;
    } on Object catch (e) {
      degraded.value = 'reopen-failed: $e';
      return false;
    }
  }

  void dispose() {
    if (_disposed) return;
    _disposed = true;
    degraded.dispose();
    recovered.dispose();
    channels.dispose();
    _frames.dispose();
    session._activeBridges.remove(this);
    session._frameRouters.remove('${info['bridgeSessionId'] ?? ''}');
  }
}

File: web/src/protocol/remoteSession.ts
/// RemoteSession + Bridge — TS 移植自 lib/protocol/remote_session.dart。
/// 信令配对层：bootstrap / workspace-list / bridge-open → rpc-frame + IPC 栈。

import { RelayClient } from './relayClient'
import { ChannelClient } from './channelClient'
import { RpcFrames, type RpcFramePayload } from './rpcFrames'
import {
  SIG_BOOTSTRAP,
  SIG_BOOTSTRAP_RESPONSE,
  SIG_BRIDGE_OPEN,
  SIG_BRIDGE_READY,
  SIG_BRIDGE_ERROR,
  SIG_VIEW_STATE_UPDATE,
  PUSH_WORKSPACE_LIST_UPDATED,
  PUSH_BRIDGE_DEGRADED,
  genId,
} from './constants'
import type { LinkParams } from './linkParams'

export type Workspace = Record<string, unknown>

export class Bridge {
  info: Record<string, unknown>
  frames!: RpcFrames
  channels!: ChannelClient
  degraded: string | null = null
  recoveredTick = 0
  private disposedFlag = false
  private degradedListeners = new Set<() => void>()

  constructor(
    public session: RemoteSession,
    info: Record<string, unknown>,
  ) {
    this.info = info
    this.attach(info)
  }

  /** 建 rpc-frame + channel 栈（构造/恢复共用）。 */
  attach(info: Record<string, unknown>): void {
    this.info = info
    this.frames = new RpcFrames({
      bridgeSessionId: String(info['bridgeSessionId'] ?? ''),
      bridgeGeneration: (info['bridgeGeneration'] as number | undefined) ?? undefined,
      recoveryId: (info['recoveryId'] as string | undefined) ?? undefined,
      send: (p) => this.session.relay.sendPayload(p),
      onMessage: (bytes) => this.channels.handleMessage(bytes),
      onLog: this.session.onLog,
    })
    this.channels = new ChannelClient((body) => this.frames.sendMessage(body), this.session.onLog)
  }

  get workspaceKey(): string | null {
    return (this.info['workspaceKey'] as string | undefined) ?? null
  }

  get initialTaskId(): string | null {
    return (this.info['initialTaskId'] as string | undefined) ?? null
  }

  get scope(): Record<string, unknown> {
    return {
      workspacePath: this.info['workspacePath'] ?? this.workspaceKey,
      ...(this.info['workspaceIdentity'] != null
        ? { workspaceIdentity: this.info['workspaceIdentity'] }
        : {}),
    }
  }

  setDegraded(reason: string | null): void {
    if (this.disposedFlag) return
    this.degraded = reason
    for (const l of this.degradedListeners) l()
  }

  onDegraded(listener: () => void): () => void {
    this.degradedListeners.add(listener)
    return () => this.degradedListeners.delete(listener)
  }

  /** 恢复重试直到桥健康（对齐 recoverWithRetry）。 */
  async recoverWithRetry(): Promise<void> {
    if (this.disposedFlag) return
    for (let attempt = 1; attempt <= 8; attempt++) {
      if (this.disposedFlag) return
      if (await this.recoverOnce()) return
      if (this.disposedFlag) return
      this.session.onLog?.(`[bridge] recovery attempt ${attempt} failed`)
      await new Promise((r) => setTimeout(r, 3000))
    }
  }

  private async recoverOnce(): Promise<boolean> {
    const key = this.workspaceKey
    if (key == null) return false
    try {
      const res = await this.session.request(
        {
          zcode_type: 'workspace-reconnect-request',
          requestId: genId('rec'),
          bridgeSessionId: this.info['bridgeSessionId'],
          workspaceKey: key,
        },
        (p) =>
          (p['zcode_type'] === 'workspace-reconnect-response' ||
            p['zcode_type'] === SIG_BRIDGE_ERROR) &&
          p['bridgeSessionId'] === this.info['bridgeSessionId'],
      )
      if (res['zcode_type'] === SIG_BRIDGE_ERROR) return false
      const info = (res['bridge'] as Record<string, unknown> | undefined) ?? {}
      if (Object.keys(info).length === 0) return false
      this.attach(info)
      this.recoveredTick++
      this.setDegraded(null)
      return true
    } catch {
      return false
    }
  }

  dispose(): void {
    this.disposedFlag = true
    this.frames.dispose()
    this.channels.dispose()
  }
}

export class RemoteSession {
  relay: RelayClient
  private payloadListeners = new Set<(p: RpcFramePayload) => void>()
  private workspaceListListeners = new Set<(result: unknown) => void>()
  private matchers = new Map<string, (p: RpcFramePayload) => boolean>()
  private completers = new Map<string, { resolve: (p: RpcFramePayload) => void; reject: (e: unknown) => void }>()
  private frameRouters = new Map<string, Bridge>()
  private pendingBridgePayloads = new Map<string, RpcFramePayload[]>()
  private bridgeGeneration = 0
  private unsubs: (() => void)[] = []
  private needsBridgeRecovery = false
  private activeBridges: Bridge[] = []

  workspaces: Workspace[] = []

  constructor(
    public params: LinkParams,
    public onLog?: (line: string) => void,
  ) {
    this.relay = new RelayClient(params, onLog)
    this.unsubs.push(
      this.relay.onPayload((p) => this.dispatch(p)),
      this.relay.onState(() => this.onRelayState()),
    )
  }

  private onRelayState(): void {
    const s = this.relay.state
    if (s === 'reconnecting' || s === 'error') {
      if (this.activeBridges.length > 0) {
        this.needsBridgeRecovery = true
        for (const b of this.activeBridges) {
          if (b.degraded == null) b.setDegraded('reconnecting')
        }
      }
      return
    }
    if (s === 'paired' && this.needsBridgeRecovery) {
      this.needsBridgeRecovery = false
      for (const b of [...this.activeBridges]) void b.recoverWithRetry()
    }
  }

  connect(): Promise<void> {
    return this.relay.start()
  }

  /** 等 relay 到 paired（对齐 waitPaired）。 */
  waitPaired(timeoutMs = 60_000): Promise<void> {
    if (this.relay.state === 'paired') return Promise.resolve()
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        off()
        reject(new Error('配对超时'))
      }, timeoutMs)
      const off = this.relay.onState(() => {
        if (this.relay.state === 'paired') {
          clearTimeout(timer)
          off()
          resolve()
        }
      })
    })
  }

  /** 信令 request：按 requestId 挂 matcher，响应不保证回带 id，人人过一遍。 */
  request(
    payload: RpcFramePayload,
    match: (p: RpcFramePayload) => boolean,
    timeoutMs = 30_000,
  ): Promise<RpcFramePayload> {
    const requestId = payload['requestId'] as string
    return new Promise((resolve, reject) => {
      this.matchers.set(requestId, match)
      this.completers.set(requestId, { resolve, reject })
      this.relay.sendPayload(payload)
      setTimeout(() => {
        if (this.completers.has(requestId)) {
          this.matchers.delete(requestId)
          this.completers.delete(requestId)
          reject(new Error(`request ${requestId} timed out`))
        }
      }, timeoutMs)
    })
  }

  private dispatch(payload: RpcFramePayload): void {
    const type = payload['zcode_type']
    if (type === PUSH_WORKSPACE_LIST_UPDATED) {
      this.workspaces = ((payload['result'] as Record<string, unknown> | undefined)?.[
        'workspaces'
      ] as Workspace[] | undefined) ?? this.workspaces
      for (const l of this.workspaceListListeners) l(payload['result'])
      return
    }
    if (type === PUSH_BRIDGE_DEGRADED) {
      const id = payload['bridgeSessionId'] as string | undefined
      const reason = String(payload['reason'] ?? 'unknown')
      this.onLog?.(`[bridge] degraded: ${id} reason=${reason}`)
      for (const b of this.activeBridges) {
        if (b.info['bridgeSessionId'] === id) {
          b.setDegraded(reason)
          void b.recoverWithRetry()
        }
      }
      return
    }
    if (type === 'rpc-frame' || type === 'rpc-frame-ack') {
      const id = payload['bridgeSessionId'] as string | undefined
      const bridge = id ? this.frameRouters.get(id) : undefined
      if (bridge) bridge.frames.accept(payload)
      else if (id) {
        const list = this.pendingBridgePayloads.get(id) ?? []
        list.push(payload)
        this.pendingBridgePayloads.set(id, list)
      }
      return
    }
    const done: string[] = []
    for (const [requestId, match] of this.matchers) {
      const completer = this.completers.get(requestId)
      if (completer && match(payload)) {
        done.push(requestId)
        completer.resolve(payload)
      }
    }
    for (const id of done) {
      this.matchers.delete(id)
      this.completers.delete(id)
    }
  }

  async bootstrap(): Promise<Record<string, unknown>> {
    const res = await this.request(
      { zcode_type: SIG_BOOTSTRAP, requestId: genId('boot') },
      (p) => p['zcode_type'] === SIG_BOOTSTRAP_RESPONSE,
    )
    return (res['result'] as Record<string, unknown>) ?? res
  }

  sendViewState(workspaceKey: string, taskId?: string): void {
    this.relay.sendPayload({
      zcode_type: SIG_VIEW_STATE_UPDATE,
      viewState: {
        activeWorkspaceKey: workspaceKey,
        ...(taskId ? { activeTaskId: taskId } : {}),
        updatedAt: Date.now(),
      },
      deviceInfo: {
        platform: 'web',
        version: this.params.appVersion ?? 'zremote',
        name: 'zremote',
      },
    })
  }

  onWorkspaceList(listener: (result: unknown) => void): () => void {
    this.workspaceListListeners.add(listener)
    return () => this.workspaceListListeners.delete(listener)
  }

  /** workspace-bridge-open → ready → rpc-frame/IPC 栈（单活动桥语义）。 */
  async openBridge(workspaceKey: string, taskId?: string, timeoutMs = 30_000): Promise<Bridge> {
    const requestedId = genId('bridge')
    const generation = ++this.bridgeGeneration
    const res = await this.request(
      {
        zcode_type: SIG_BRIDGE_OPEN,
        requestId: genId('open'),
        bridgeSessionId: requestedId,
        bridgeGeneration: generation,
        workspaceKey,
        ...(taskId ? { taskId } : {}),
      },
      (p) =>
        (p['zcode_type'] === SIG_BRIDGE_READY || p['zcode_type'] === SIG_BRIDGE_ERROR) &&
        p['bridgeSessionId'] === requestedId,
      timeoutMs,
    )
    if (res['zcode_type'] === SIG_BRIDGE_ERROR) {
      throw new Error(`workspace-bridge-error: ${res['error'] ?? JSON.stringify(res)}`)
    }
    const info = (res['bridge'] as Record<string, unknown> | undefined) ?? {}
    this.onLog?.(`[bridge] ready: ${JSON.stringify(info)}`)
    const bridge = new Bridge(this, info)
    this.attachStack(bridge, info)
    this.activeBridges.push(bridge)
    const key = (info['workspaceKey'] as string | undefined) ?? workspaceKey
    this.sendViewState(key, bridge.initialTaskId ?? taskId)
    return bridge
  }

  private attachStack(bridge: Bridge, info: Record<string, unknown>): void {
    bridge.attach(info)
    const id = String(info['bridgeSessionId'] ?? '')
    this.frameRouters.set(id, bridge)
    const pending = this.pendingBridgePayloads.get(id)
    if (pending) {
      this.pendingBridgePayloads.delete(id)
      for (const payload of pending) bridge.frames.accept(payload)
    }
  }

  dispose(): void {
    for (const u of this.unsubs) u()
    for (const b of [...this.activeBridges]) b.dispose()
    this.relay.dispose()
    this.payloadListeners.clear()
  }
}
